rfcs/0050-content-addressed-paths.md

```diff
@@ -0,0 +1,261 @@
+on-disk)
+
+### Building a ca derivation
+
+ca derivations are derivations with the `contentAdressed` argument set to
+`true`.
+
+The process for building a content-adressed derivation is the following:
+- We build it like a normal derivation to get an output path `$out`.
+- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
+- We move `$out` to `/nix/store/$chash-$name`
+- We create an alias path from `$out` to `/nix/store/$chash-$name`
+
+[^modulo-hashing]: We can possibly normalize all the self-references before
+  computing the hash and rewrite them when moving the path to handle paths with
+  self-references, but this isn't strictly required for a first iteration
+
+### Building a normal derivation
+
+The process for building a normal derivation is the following:
+- We look into the drv for all the inputs paths of the build
+- For each input path, we look whether the path is an alias. If so we replace it
+  by its target
+- We compute the `dynamic` output of the derivation from the patched version
+- We then try to substitute and build the new derivation
+- We create an alias path from the `static` output to the `dynamic` one
+
+## Wrapping it up
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Obviously, this makes the Nix model more complicated than what it is now.  In
+  particular, the caching model needs some modifications (see [caching]);
+
+- We specify that only a sub-category of derivations can safely be marked as
+  `contentAdressed`, but there's no way to enforce these restricitions;
+
+- This will probably be a breaking-change for some tooling since the output path
+  that's stored in the `.drv` files doesn't correspond to the actual on-disk
+  path the output will be stored in (because it might just be an alias for the
+  other path)
+
+# Alternatives
+[alternatives]: #alternatives
+
+[RFC 0017][] is another proposal with the
+same end-goal. The big difference between these two is in the scope they cover:
+RFC 0017 is about fundamentally changing the base model of Nix, while this
+proposal suggests to make only the minimal amount of changes to the current
+model to allow the content-adressed model to live in parallel (which would open
+the way to a fully content-adressed store as RFC0017, but in a much more
+incremental way).
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Caching
+[caching]: #caching
+
+The big unresolved question is about the caching of content-adressed paths.
+As [Eelco's phd thesis][NixPhd] states it, caching ca paths raises a number of
+questions when building that path is non-deterministic (because two different
+stores can have two different outputs for the same path, which might lead to
+some dependencies being duplicated in the closure of a dependency).
+There exist some solutions to this problem (including one presented in Eelco's
+thesis), but for the sake of simplicity, this RFC simply forbids to mark a
+derivation as ca if its build is not deterministic.
```

Comment:
Is `if its build is not deterministic` == FOD here?

If not, how do you assume a build is not deterministic?

Reply:
That depends on how we end-up implementing it, but it's reasonably simple to implement something like "your local build of `foo` differs from the one on the remote cache => You get cache misses for everything that depends on `foo`" (so some form of local-only FOD)